include/nlohmann/detail/iterators/iteration_proxy.hpp

```diff
@@ -99,11 +99,38 @@ template<typename IteratorType> class iteration_proxy
         {
             return anchor.value();
         }
+
+        /// The get function is required to add support for Structued Bindings
+        /// Structured Bindings will help simplify the way code is written
+        /// For these functions
+        template<std::size_t N, typename = typename std::enable_if<N >= 2>::type >
+        void get()const
+        {
+            static_assert(N >= 2, "Has Max 2 Elements:- Key or Value");
+        }
+        template<std::size_t N, typename = typename  std::enable_if<N == 0>::type>
+        auto get() const -> decltype(key())
+        {
+            return key();
+        }
+        template<std::size_t N, typename = typename std::enable_if<N == 1>::type>
+        auto get() const  -> decltype(value())
+        {
+            return value();
+        }
```

Comment:
I think it's better to make those non-member, to avoid users calling them (not even talking about the possible confusion with `json::get`)

Reply:
That might not work.
Because the basic idea with Structured Bindings is that they return key and value
Key at Index 0
And Value at Index 1
At least when used with items.
Making them public/anything else would render them incompatible with Structured Bindings support in json.items

The issue was with iteration_proxy_internal being declared as private however, I have made it public so it should be compatible now

Reply:
About non-members, the prototype is the same, but it takes an argument, where the member-function takes none. It should work the same either way.

Reply:
Not really sure.
Because as far as I know, the way Structured Bindings work is that they consider everything is a Tuple/Tuple Element.

For example Structured Binding related to the code

```
   for (auto const& [key,value] : json.items())
   {
      std::cout << "key: " << key << ", value: " << value << '\n';
   }

```
Is probably same as
```
   for (auto const& elem : json.items())
   {
      auto const key = elem.get<0>();
      auto const value = elem.get<1>();
      std::cout << "key: " << key << ", value: " << value << '\n';
   }
```

Reply:
From [cppreference](https://en.cppreference.com/w/cpp/language/structured_binding):

* `e.get<i>()`, if lookup for the identifier get in the scope of `E` by class member access lookup finds at least one declaration that is a function template whose first template parameter is a non-type parameter.

* Otherwise, `get<i>(e)`, where `get` is looked up by argument-dependent lookup only, ignoring non-ADL lookup.

Your assumption is correct when the first bullet point is matched, but when the second one is matched, code is equivalent to:

```cpp
for (auto const& elem : json.items())
{
    auto const key = std::get<0>(elem);
    auto const value = std::get<1>(elem);
    std::cout << "key: " << key << ", value: " << value << '\n';
}
```

Reply:
> 
> 
> In generic context, one would always use `std::get<0>(...)`, especially to avoid writing `j.template get<0>()`

I apologise.
It seems I can push it out to std namespace
The code would look similar to

```
namespace std {

   // Required for Structured Bindings support
   template<std::size_t N, typename = typename std::enable_if<N >= 2>::type>
   void get(nlohmann::detail::iteration_proxy<nlohmann::json::iterator>::iteration_proxy_internal&)
   {
      static_assert(N >= 2, "Has Max 2 Elements:- Key or Value");
   }
   template<std::size_t N, typename = typename  std::enable_if<N == 0>::type>
   auto get(nlohmann::detail::iteration_proxy<nlohmann::json::iterator>::iteration_proxy_internal const& x)  -> decltype(x.key())
   {
      return x.key();
   }
   template<std::size_t N, typename = typename std::enable_if<N == 1>::type>
   auto get(nlohmann::detail::iteration_proxy<nlohmann::json::iterator>::iteration_proxy_internal const& x)   -> decltype(x.value())
   {
      return x.value();
   }

   template<std::size_t N>
   struct tuple_element<N, nlohmann::detail::iteration_proxy<nlohmann::json::iterator>::iteration_proxy_internal> {
      using type = decltype(get<N>(std::declval<nlohmann::detail::iteration_proxy<nlohmann::json::iterator>::iteration_proxy_internal const&>()));
   };

   template<>
   struct tuple_size<nlohmann::detail::iteration_proxy<nlohmann::json::iterator>::iteration_proxy_internal>
      : std::integral_constant<std::size_t, 2> {};
}
```

We still however would require iteration_proxy_internal to be public and json::iterator to be a parameter which doesn't seem to be so bad.

Reply:
What is the compiler error? Could you show the full snippet?

Reply:
error C2672: 'get': no matching overloaded function found
  C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\include\utility(482): note: see reference to class template instantiation 'std::tuple_element<0,_Tuple>' being compiled
          with
          [
              _Tuple=nlohmann::detail::iteration_proxy<nlohmann::detail::iter_impl<nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer>>>::iteration_proxy_internal
          ]
JSONTest.cpp(10): note: see reference to class template instantiation 'std::tuple_element<0,const nlohmann::detail::iteration_proxy<nlohmann::detail::iter_impl<nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer>>>::iteration_proxy_internal>' being compiled
include\nlohmann\json.hpp(20236): error C2784: 'tuple_element<_Idx,std::pair<_Ty1,_Ty2>>::type &std::get(std::pair<_Ty1,_Ty2> &) noexcept': could not deduce template argument for 'std::pair<_Ty1,_Ty2> &' from 'const nlohmann::detail::iteration_proxy<nlohmann::detail::iter_impl<nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer>>>::iteration_proxy_internal'
  C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\include\utility(600): note: see declaration of 'std::get'

The error occurred in my test project

Reply:
Indeed, I forgot about that one, I guess you have to forward declare json::iterator then...

Reply:
Nope the new commit fails.
get function probably can't properly lie within the std namespace
It fails with an error that it was unable to find get when a Structured Binding is used

Reply:
I think we should just stick to the old commit that placed the get function in iterator_proxy

Reply:
I will take a look tomorrow, I think we're close

Reply:
Auto Deduction of Return Types also fails on a few older compilers so decided to copy paste return types there from the original method.
The issue with this is that if key/value's types are changed, then get's types need to be changed manually
https://travis-ci.org/nlohmann/json/jobs/466093802